lib/ruby_smb/dispatcher/socket.rb

```diff
@@ -41,11 +49,15 @@ def send_packet(packet)
       # which are assumed to be the NetBiosSessionService header.
       # @return [String]
       def recv_packet
-        IO.select([@tcp_socket])
+        if IO.select([@tcp_socket], nil, nil, @read_timeout).nil?
+          raise RubySMB::Error::CommunicationError, "Read timeout expired when reading from the Socket (timeout=#{@read_timeout})"
+        end
         nbss_header = @tcp_socket.read(4) # Length of NBSS header. TODO: remove to a constant
-        raise ::RubySMB::Error::NetBiosSessionService, 'NBSS Header is missing' if nbss_header.nil?
+        raise ::RubySMB::Error::NetBiosSessionService, 'NBSS Header is missing' if nbss_header.nil? || nbss_header.empty?
```

Comment:
I had issues with some specific Socket implementations that return an empty string instead of nil when something wrong happens.